evap/results/tools.py

```diff
@@ -41,28 +41,43 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
         self.visible_by_delegation_count = visible_by_delegation_count
 
 
+def create_rating_result(question, answer_counters, additional_text_result=None):
+    if answer_counters is None:
+        return RatingResult(question, answer_counters, additional_text_result)
+    if any(counter.count != 0 for counter in answer_counters):
+        return AnsweredRatingResult(question, answer_counters, additional_text_result)
+    return PublishedRatingResult(question, answer_counters, additional_text_result)
```

Comment:
I wonder if these invariants would be better kept in the `__init__` of the respective classes. And then, we could `try` to make the most concrete one first, but I am not sure if it would be nicer. But we could add a helper to `PublishedRatingResult` (which the child class would also inherit):
```python
@classmethod
def try_to_create(cls, *args, **kwargs) -> Optional[Self]:
    try:
        return cls(*args, **kwargs)
    except ValueError:
        return None
```
Or maybe it's too weird. @richardebeling ?

Reply:
I think the current code is fine -- your suggestion also would still need the factory method, right? In that case, I'd try to stick to the simpler code (which I think is the current code)